### ISSUE TITLE
Not displaying "&" in password

### ISSUE THREAD

**Issue by britmap** [darkarnium/secpub#6]
Whilst running the PoC against my vulnerable WNDR3700v4 and WNR1000v2 routers, I noticed that the Admin Password reported wasn't always correct. My password happens to include an "&" ampersand character so I wonder if that is the issue?
For example setting the WNDR3700v4 admin password to "qwerty&123" the PoC code reports just "qwerty123". Or setting the admin password to "abcd&DEFG" the PoC reports just "abcd".
Regards,
Chris

**Comment by darkarnium** [comment]
Hi there,

It looks like the SOAP service on the NetGear devices aren't XML encoding the values before inserting them into the document as part of the response - which seems to be a breach of the XML standard:
- http://www.w3.org/TR/2000/REC-xml-20001006#syntax

As a result of this lack of encoding, the parser is failing to parse the string from that node. I will see if I can patch around it.

Cheers for the bug report!

**Comment by darkarnium** [comment]
Hey there,

TL;DR: It looks like the code generating the XML response documents on these devices is not encoding data per the standard.

In order to fix the failed rendering of '&'  I can do a VERY hacky `gsub()` on the output from `response.body()` - in order to replace any occurrence of the ampersand with the correct XML encoded value. That said, I'm disinclined to do this as if the document were to occur any correctly encoded value, it would be horribly broken by this hack.

I've also done a bit more testing and found that other characters that _SHOULD_ be XML encoded aren't being encoded at all. This causes headaches as the XML documents are completely invalid when certain un-encoded characters are written into the document body:
- `<` as `&lt;`
- `>` as `&gt;`
- `&` as `&amp;`
- `"` as `&quot;`

As you can imagine, I can't `gsub()` characters which are used to define XML nodes :)

**Comment by britmap** [comment]
So a vulnerability in Netgear code reveals another bug in Netgear code? :) Thanks for looking.

**Comment by darkarnium** [comment]
It would appear that way, thanks for reporting it! :)

Cheers,
Peter
